tests/src/test_kahan_summation.cpp

```diff
@@ -0,0 +1,182 @@
+    }
+}
+
+TEST_CASE("logsumexp_sort_kahan_inplace", "[kahan_summation]") {
+    const int len = 5;
+    double neg_inf = -std::numeric_limits<double>::infinity();
+    double arr[len] = {neg_inf, neg_inf, neg_inf, neg_inf, neg_inf};
+
+    SECTION("-inf array returns -inf - overload (begin, end)") {
+        double res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, arr + len);
+        REQUIRE(res == -std::numeric_limits<double>::infinity());
+        res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, len);
+        REQUIRE(res == -std::numeric_limits<double>::infinity());
+    }SECTION("zero length array yields -inf") {
+        arr[0] = 3.;
+        arr[1] = 5.;
+        double res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, arr);
+        REQUIRE(res == -std::numeric_limits<double>::infinity());
+
+        res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, 0);
+        REQUIRE(res == -std::numeric_limits<double>::infinity());
+    }SECTION("Kahan logsumexp sort inplace yields correct value - low to high") {
+        arr[2] = 3.;
+        arr[4] = 100.;
+
+        SECTION("overload (begin, end)") {
+            double res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, arr + 5);
+            REQUIRE(res == std::log(std::exp(3.) + std::exp(100.)));
+        }SECTION("overload (begin, size)") {
+            double res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, 5);
+            REQUIRE(res == std::log(std::exp(3.) + std::exp(100.)));
+        }
+    }SECTION("Kahan logsumexp sort inplace yields correct value - high to low") {
+        arr[4] = 3.;
+        arr[2] = 100.;
+        SECTION("overload (begin, end)") {
+            double res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, arr + 5);
+            REQUIRE(res == std::log(std::exp(3.) + std::exp(100.)));
+        }SECTION("overload (begin, size)") {
+            double res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, 5);
+            REQUIRE(res == std::log(std::exp(3.) + std::exp(100.)));
+        }
+    }
+    SECTION("Kahan logsumexp sort inplace does not overflow") {
+        arr[4] = 3.;
+        arr[2] = 10.e8;
+        SECTION("overload (begin, end)") {
+            REQUIRE(std::log(std::exp(arr[4]) + std::exp(arr[2])) == std::numeric_limits<double>::infinity());
+            double res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, arr + 5);
+            REQUIRE(res < std::numeric_limits<double>::infinity());
+        }SECTION("overload (begin, size)") {
+            REQUIRE(std::log(std::exp(arr[4]) + std::exp(arr[2])) == std::numeric_limits<double>::infinity());
+            double res = deeptime::numeric::kahan::logsumexp_sort_kahan_inplace(arr, 5);
+            REQUIRE(res < std::numeric_limits<double>::infinity());
+        }
+    }
+}
+
+
+//struct mock_np_array : np_array_nfc<int> {
```

Comment:
this is actually part of a larger todo where all of deeptime c++ code is refactored so that it can be used with any api-compliant container instead of just numpy arrays, perhaps we should remove this comment and add an issue